web/src/information_density.ts

```diff
@@ -171,3 +198,122 @@ export function initialize(): void {
+    "web_font_size_px",
+    "web_line_height_percent",
+]);
+
+export function enable_or_disable_control_buttons($container: JQuery): void {
+    const info_density_properties = z
+        .array(information_density_properties_schema)
+        .parse(["web_font_size_px", "web_line_height_percent"]);
+    for (const property of info_density_properties) {
+        const $button_group = $container.find(`[data-property='${CSS.escape(property)}']`);
+        const $current_elem = $button_group.find(".current-value");
+        const current_value = Number.parseInt($current_elem.text(), 10);
+
+        $button_group
+            .find(".default-button")
+            .prop("disabled", current_value === INFO_DENSITY_VALUES_DICT[property].default);
+        $button_group
+            .find(".increase-button")
+            .prop("disabled", current_value === INFO_DENSITY_VALUES_DICT[property].maximum);
+        $button_group
+            .find(".decrease-button")
+            .prop("disabled", current_value === INFO_DENSITY_VALUES_DICT[property].minimum);
+    }
+}
+
+export function get_new_value_for_information_density_settings(
+    $elem: JQuery,
+    changed_property: "web_font_size_px" | "web_line_height_percent",
+): number {
+    const $current_elem = $elem.closest(".button-group").find(".current-value");
+    const current_value = Number.parseInt($current_elem.text(), 10);
+
+    let new_value;
+    if ($elem.hasClass("default-button")) {
+        new_value = INFO_DENSITY_VALUES_DICT[changed_property].default;
+    } else if ($elem.hasClass("increase-button")) {
+        new_value = current_value + INFO_DENSITY_VALUES_DICT[changed_property].step_value;
+    } else {
+        new_value = current_value - INFO_DENSITY_VALUES_DICT[changed_property].step_value;
+    }
+    $current_elem.text(current_value);
+
+    return new_value;
+}
+
+export function update_information_density_settings(
+    $elem: JQuery,
+    changed_property: "web_font_size_px" | "web_line_height_percent",
+    new_value: number,
+): {
+    web_font_size_px?: number;
+    web_line_height_percent?: number;
+    dense_mode?: boolean;
+} {
+    const changed_settings: {
+        web_font_size_px?: number;
+        web_line_height_percent?: number;
+        dense_mode?: boolean;
+    } = {};
+    changed_settings[changed_property] = new_value;
+
+    user_settings[changed_property] = new_value;
+    $elem.closest(".button-group").find(".current-value").text(new_value);
+
+    set_base_typography_css_variables();
+    calculate_timestamp_widths();
+
+    if (
+        user_settings.dense_mode &&
+        (user_settings.web_font_size_px !== LEGACY_FONT_SIZE_PX ||
+            user_settings.web_line_height_percent !== LEGACY_LINE_HEIGHT_PERCENT)
+    ) {
+        user_settings.dense_mode = false;
+        changed_settings.dense_mode = false;
+        $("body").toggleClass("less-dense-mode");
+        $("body").toggleClass("more-dense-mode");
+    }
```

Comment:
Remind me what `dense_mode` does mechanically? This seems like it drops dense mode, but will never restore it, and I'm not sure if that's intended.

Reply:
This was incorrect. Updated to correctly add and remove the required classes.

Also, currently we only change `dense_mode` if and we set it to False -
- `dense_mode` is set to `True` originally.
- New values of font size is not 14 and of line height is not 122.

But we do not set it to `True` again if font size is 14 and line height is 122. This is same as was the behavior previously.